### PR TITLE
feat(LDAP): implement IIsAdmin interface

### LINKED ISSUE #6428: Give admin rights to a specific LDAP group
My users are stored in an LDAP directory and I want to give admin permissions to some of them. I could manually put them in the default admin group of nextcloud but it would be more efficient to be able to select a group from LDAP as admin group. Could that be implemented?

### PR DESCRIPTION
**Pull request by blizzz**
* Resolves: #6428 

## Summary

Promotes an LDAP group (per LDAP configuration) to an admin group.

A group can either be promoted via a dedicates occ call whres the group parameter can be a nextcloud group ID or a group name that will be search against – an exact match is required in that case.:

```
$ php occ ldap:promote-group  --help 2>/dev/null
Description:
  declares the specified group as admin group (only one is possible per LDAP configuration)

Usage:
  ldap:promote-group [options] [--] <group>

Arguments:
  group                 the group ID in Nextcloud or a group name

Options:
  -y, --yes             do not ask for confirmation
…

# Example
$ php occ ldap:promote-group  "Nextcloud Admins"
Promote Nextcloud Admins to the admin group (y|N)? y
Group Nextcloud Admins was promoted

$ php occ ldap:promote-group  "Paramount Court"
Promote Nextcloud Admins to the admin group and demote Nextcloud Admins (Group ID: nextcloud_admins) (y|N)? y
Group Paramount Court was promoted

$ php occ ldap:promote-group  "Paramount Court"
The specified group is already promoted
```

It is also possible to set the admin group mapping using `occ ldap:set-config $configId ldapAdminGroup $groupId` but as the Nextcloud group ID might not be known (yet) it is especially recommnded for automatized setups to use this command, that would also pull in the group and determine the group ID.

In order to demote or reset a promotion, an empty string should be set against to the targeted config's ldapAdminGroup:

```
# Reset an admin group mapping via set-config
occ ldap:set-config $configId ldapAdminGroup ""
# Example
occ ldap:set-config s01 ldapAdminGroup ""
```

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [X] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [X] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Comment by blizzz** [comment]
Need to check the integration tests, the positive one is failing :thinking: 

User should be in the target group
https://github.com/nextcloud/docker-ci/blob/master/openldap/LDIFs/ordinary_small.ldif#L160-L163

File: apps/user_ldap/lib/Command/PromoteGroup.php [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+	protected function configure(): void {
+		$this
+			->setName('ldap:promote-group')
+			->setDescription('declares the specified group as admin group (only one is possible per LDAP configuration)')
+			->addArgument(
+				'group',
+				InputArgument::REQUIRED,
+				'the group ID in Nextcloud or a group name'
+			)
+			->addOption(
```
Comment (juliushaertl) [inline apps/user_ldap/lib/Command/PromoteGroup.php:61]: Optional: We could have a option to remove the group again or set an empty config.

Comment (blizzz) [reply apps/user_ldap/lib/Command/PromoteGroup.php:61]: That is what slipped my mind yesterday, yes.

Comment (blizzz) [reply apps/user_ldap/lib/Command/PromoteGroup.php:61]: The tricky part here is that we do not specify the config prefix, but detect it depending on the group ownership. So, only resetting any group would need the config prefix. 

Resetting by a group name or id could fail if the group is deleted, but we may try to guess by comparing stored config – not full proof. I tend to go with documenting how to do with with set-config instead (does not require another command).

Apropos group deletion, this is something we could take into account as well and remove the promotion when a group disappears – although this is a bit atypical from what we usually do (like keeping users in place), but actually safer than keeping the entry. Questionable upon unassigned the group id (clearing the mappings), which should not be done in prod though (would keep it in that case).

**Comment by blizzz** [comment]
> Apropos group deletion, this is something we could take into account as well and remove the promotion when a group disappears – although this is a bit atypical from what we usually do (like keeping users in place), but actually safer than keeping the entry. Questionable upon unassigned the group id (clearing the mappings), which should not be done in prod though (would keep it in that case).

An addition for later for we do not delete groups yet.

**Comment by blizzz** [comment]
> Resetting by a group name or id could fail if the group is deleted, but we may try to guess by comparing stored config – not full proof. I tend to go with documenting how to do with with set-config instead (does not require another command).

Updated the description, that will go into documentation eventually.

File: apps/user_ldap/lib/Command/PromoteGroup.php (hunk 1 above)
Comment (juliushaertl) [reply apps/user_ldap/lib/Command/PromoteGroup.php:56]: Ah ok, I misread the changeset then in regards how this works with multiple ldap configurations. Thanks for explaining, makes sense then.

**Review by juliusknorr — APPROVED** [review]

File: apps/user_ldap/lib/Command/PromoteGroup.php [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	}
+
+	protected function configure(): void {
+		$this
+			->setName('ldap:promote-group')
+			->setDescription('declares the specified group as admin group (only one is possible per LDAP configuration)')
+			->addArgument(
+				'group',
+				InputArgument::REQUIRED,
+				'the group ID in Nextcloud or a group name'
```
Comment (come-nc) [inline apps/user_ldap/lib/Command/PromoteGroup.php:54]: LDAP dn does not work?

File: apps/user_ldap/lib/Command/PromoteGroup.php [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+		if ($input->getOption('yes') === false) {
+			/** @var QuestionHelper $helper */
+			$helper = $this->getHelper('question');
+
+			$idLabel = '';
+			if ($group->getGID() !== $group->getDisplayName()) {
+				$idLabel = sprintf(' (Group ID: %s)', $group->getGID());
+			}
+
+			$q = new Question(sprintf('Promote %s%s to the admin group (y|N)? ', $group->getDisplayName(), $idLabel));
```
Comment (come-nc) [inline apps/user_ldap/lib/Command/PromoteGroup.php:74]: The question should also inform if this is unpromoting another group which was previously set

File: apps/user_ldap/lib/Command/PromoteGroup.php (hunk 2 above)
Comment (blizzz) [reply apps/user_ldap/lib/Command/PromoteGroup.php:54]: not unless you specific it in the search attributes

File: apps/user_ldap/lib/Command/PromoteGroup.php (hunk 3 above)
Comment (blizzz) [reply apps/user_ldap/lib/Command/PromoteGroup.php]: done, added a commit with this handling

**Review by come-nc — APPROVED** [review]

**Comment by blizzz** [comment]
addressed the php-cs test complaint.

**Pull request merged**